pkg/controller/volume/attachdetach/populator/desired_state_of_world_populator.go

```diff
@@ -113,3 +122,208 @@ func (dswp *desiredStateOfWorldPopulator) findAndRemoveDeletedPods() {
+						pod.Name,
+						err)
+					continue
+				}
+
+				// check if the pod and its volume is added to dsw
+				added, err := dswp.isPodVolumePreviouslyAdded(uniquePodName, volumeSpec, nodeName)
+				if err != nil {
+					glog.V(10).Infof("Check whether pod references the specified volume is added to dsw falied: %v", err)
+					continue
+				} else if added {
+					glog.V(10).Infof("Pod references the specified volume is already added to dsw")
+					continue
+				}
+
+				dswp.desiredStateOfWorld.AddPod(uniquePodName, pod, volumeSpec, nodeName)
+			}
+		}
+	}
+
+}
+
+// createVolumeSpec creates and returns a volume.Spec object for the
+// specified volume. It dereference any PVC to get PV objects, if needed.
+func (dswp *desiredStateOfWorldPopulator) createVolumeSpec(podVolume v1.Volume, podNamespace string) (*volume.Spec, error) {
+	if pvcSource :=
+		podVolume.VolumeSource.PersistentVolumeClaim; pvcSource != nil {
+		glog.V(10).Infof(
+			"Found PVC, ClaimName: %q/%q",
+			podNamespace,
+			pvcSource.ClaimName)
+
+		// If podVolume is a PVC, fetch the real PV behind the claim
+		pvName, pvcUID, err := dswp.getPVCExtractPV(
+			podNamespace, pvcSource.ClaimName)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"error processing PVC %q/%q: %v",
+				podNamespace,
+				pvcSource.ClaimName,
+				err)
+		}
+
+		glog.V(10).Infof(
+			"Found bound PV for PVC (ClaimName %q/%q pvcUID %v): pvName=%q",
+			podNamespace,
+			pvcSource.ClaimName,
+			pvcUID,
+			pvName)
+
+		// Fetch actual PV object
+		volumeSpec, err :=
+			dswp.getPVSpec(pvName, pvcSource.ReadOnly, pvcUID)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"error processing PVC %q/%q: %v",
+				podNamespace,
+				pvcSource.ClaimName,
+				err)
+		}
+
+		glog.V(10).Infof(
+			"Extracted volumeSpec (%v) from bound PV (pvName %q) and PVC (ClaimName %q/%q pvcUID %v)",
+			volumeSpec.Name,
+			pvName,
+			podNamespace,
+			pvcSource.ClaimName,
+			pvcUID)
+
+		return volumeSpec, nil
+	}
+
+	// Do not return the original volume object, since the source could mutate it
+	clonedPodVolumeObj, err := api.Scheme.DeepCopy(&podVolume)
+	if err != nil || clonedPodVolumeObj == nil {
+		return nil, fmt.Errorf(
+			"failed to deep copy %q volume object. err=%v", podVolume.Name, err)
+	}
+
+	clonedPodVolume, ok := clonedPodVolumeObj.(*v1.Volume)
+	if !ok {
+		return nil, fmt.Errorf(
+			"failed to cast clonedPodVolume %#v to v1.Volume",
+			clonedPodVolumeObj)
+	}
+
+	return volume.NewSpecFromVolume(clonedPodVolume), nil
+}
+
+// getPVCExtractPV fetches the PVC object with the given namespace and name from
+// the API server extracts the name of the PV it is pointing to and returns it.
+// An error is returned if the PVC object's phase is not "Bound".
+func (dswp *desiredStateOfWorldPopulator) getPVCExtractPV(namespace string, claimName string) (string, types.UID, error) {
+	pvc, err :=
+		dswp.kubeClient.Core().PersistentVolumeClaims(namespace).Get(claimName, metav1.GetOptions{})
```

Comment:
Instead of using `kubeClient` to fetch a fresh PVC here and PV below, how about passing in and using the pv and pvc informers? That way we can avoid increasing the load on the API server.